state/application.go

```diff
@@ -640,29 +654,98 @@ func (a *Application) IsExposed() bool {
 	return a.doc.Exposed
 }
 
+// ExposedEndpoints returns a list of explicitly selected endpoints for this
+// application that should be reachable once the application is exposed. An
+// empty returned list means that all endpoints should be reachable.
+func (a *Application) ExposedEndpoints() []string {
+	return a.doc.ExposedEndpoints
+}
+
+// ExposeToSpaceIDs returns a list of space IDs that should be able to reach
+// this application's ports when the application is exposed.
+func (a *Application) ExposeToSpaceIDs() []string {
+	return a.doc.ExposeToSpaceIDs
+}
+
+// ExposeToCIDRs returns a list of CIDRs that should be able to reach this
+// application's ports when the application is exposed.
+func (a *Application) ExposeToCIDRs() []string {
+	return a.doc.ExposeToCIDRs
+}
+
 // SetExposed marks the application as exposed.
 // See ClearExposed and IsExposed.
-func (a *Application) SetExposed() error {
-	return a.setExposed(true)
+func (a *Application) SetExposed(endpoints, toSpaceIDs, toCIDRs []string) error {
+	if len(endpoints) > 0 {
+		bindings, _, err := readEndpointBindings(a.st, a.globalKey())
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		for _, endpoint := range endpoints {
+			if _, found := bindings[endpoint]; !found {
+				return errors.NotFoundf("endpoint %q", endpoint)
+			}
+		}
+	}
+
+	if len(toSpaceIDs) != 0 {
+		allSpaceInfos, err := a.st.AllSpaceInfos()
```

Comment:
Nit:
```go
if len(endpoints) > 0
...
if len(toSpaceIDs) != 0
```

Reply:
The latter actually compiles to faster code ;-)